utils/forms/forms.js

```diff
@@ -0,0 +1,368 @@
+      }
+
+      if (digitsOnly.length > 10) {
+        errorMessages.push('There should only be 10 digits in this entry.');
+      }
+    }
+  });
+
+  // Set custom validity if there are any errors
+  if (errorMessages.length > 0) {
+    isValid = false;
+    input.setCustomValidity(errorMessages.join('\n')); // This will trigger :invalid
+    showError(input, errorMessages);
+  } else {
+    clearError(input);
+  }
+
+  return isValid;
+}
+
+function validateForm(form) {
+  let isValid = true;
+
+  // Loop through all form inputs
+  const formInputs = Array.from(form.elements);
+  formInputs.forEach((input) => {
+    // Skip non-input elements like buttons
+    if (input.type !== 'submit' && input.type !== 'button') {
+      const inputValid = validateInput(input);
+      if (!inputValid) {
+        isValid = false; // Mark the form as invalid if any field fails
+      }
+    }
+  });
+
+  return isValid;
+}
+
+function buildInput(field) {
+  const input = document.createElement('input');
+  input.type = field.type || 'text';
+  input.name = field.name || '';
+  input.placeholder = field.placeholder || '';
+
+  if (field.value !== undefined && field.value !== null) {
+    input.value = field.value;
+  }
+
+  if (field.min !== undefined && field.min !== null) {
+    input.min = field.min;
+  }
+
+  // Apply additional properties if they exist
+  if (field.required) input.required = true;
+  if (field.minLength) input.minLength = field.minLength;
+  if (field.maxLength) input.maxLength = field.maxLength;
+  if (field.max) input.max = field.max;
+
+  // Add validation rules as a data attribute
+  if (field.validation && Array.isArray(field.validation)) {
+    input.dataset.validation = JSON.stringify(field.validation);
+  }
+
+  input.addEventListener('change', () => {
+    input.classList.add('touched');
+    validateInput(input);
+  });
+
+  return input;
+}
+
+function buildTextArea(field) {
+  const textarea = document.createElement('textarea');
+  textarea.name = field.name || ''; // Sets name attribute
+  textarea.placeholder = field.placeholder || ''; // Sets placeholder if provided
+  textarea.value = field.value || ''; // Sets default value if provided
+
+  // Apply additional properties if they exist
+  if (field.required) textarea.required = true;
+  if (field.rows) textarea.rows = field.rows; // Number of visible rows
+  if (field.cols) textarea.cols = field.cols; // Number of visible columns
+  if (field.maxLength) textarea.maxLength = field.maxLength; // Max length for input
+
+  textarea.addEventListener('change', () => {
+    textarea.classList.add('touched');
+    validateInput(textarea, field);
+  });
+
+  return textarea;
+}
+
+function buildSelect(field) {
+  const select = document.createElement('select');
+  select.name = field.name || ''; // Sets name attribute
+
+  if (field.required) select.required = field.required;
+  if (field.value) select.value = field.value;
```

Comment:
i don't think a value can be set before options are added? and we should check to make sure the value actually exists in options.